### PR TITLE
Deferred script loading v2

### PR DESCRIPTION
**Pull request by michaelletzgus**
Slightly improved version of  https://github.com/nextcloud/server/pull/3656

Problem before this PR:

`<script>` loading was performed asynchronously, loading 50-100 external js scripts an executing them sequentially was slow.

Example:
Full reload of "Files" app with almost all AddOns activated lasts 12.1 sec, loading 125 external .js.

Full discussion: #2272

Fix provided by this PR:

Script loading is done deferred with `<script defer>`.
That means:
* Script download starts as soon as the parser reaches the script tag
* Scripts will be downloaded simultaneously 
* Script execution is performed in order(!) after parsing is finished

Modifications:

Templates:
Placed loading of all external resources close together without any interruption by inline script or other code. The inline script code is placed last in the document - but will be executed first(!) because all other external script are deferred (see above).

Unnecessary `type=` has been removed from inline `<script>`

template.php:
`<script>` tags injected by AddOns take the 'defer' attributes here, but only when `src` attribute is present, otherwise `defer` is not allowed.

`<script defer>`:
https://www.w3schools.com/tags/att_script_defer.asp
https://developer.mozilla.org/de/docs/Web/HTML/Element/script
https://varvy.com/pagespeed/defer-loading-javascript.html
http://www.growingwiththeweb.com/2014/02/async-vs-defer-attributes.html
http://peter.sh/experiments/asynchronous-and-deferred-javascript-execution-explained/
https://www.html5rocks.com/en/tutorials/speed/script-loading/

New behavior:
Full (re)loading of NC webpages is now 3-10 times faster, depending on link latency and the underlying protocol (http1, http2) .
When using http2 (or http1 pipelining...) huge amounts of external resource should not be a problem any more.  

=> Full reload of "Files" app with almost all AddOns activated now lasts 2.3 sec, loading 125 external .js.

Possible drawback of this solution:
IE<=9 does not support `defer` properly, order of execution is NOT guaranteed. This may cause problems with script dependencies!

### CODE REVIEW

File: core/templates/layout.base.php [hunk 1]
```diff
@@ -18,15 +18,15 @@
 		<?php foreach($_['printcssfiles'] as $cssfile): ?>
 			<link rel="stylesheet" href="<?php print_unescaped($cssfile); ?>" media="print">
 		<?php endforeach; ?>
+		<?php foreach ($_['jsfiles'] as $jsfile): ?>
+			<script defer nonce="<?php p(\OC::$server->getContentSecurityPolicyNonceManager()->getNonce()) ?>" src="<?php print_unescaped($jsfile); ?>"></script>
+		<?php endforeach; ?>
+		<?php print_unescaped($_['headers']); ?>
 		<?php if (isset($_['inline_ocjs'])): ?>
-			<script nonce="<?php p(\OC::$server->getContentSecurityPolicyNonceManager()->getNonce()) ?>" type="text/javascript">
+			<script nonce="<?php p(\OC::$server->getContentSecurityPolicyNonceManager()->getNonce()) ?>">
```
Comment (nickvergessen) [inline core/templates/layout.base.php]: is this change required?

Comment (michaelletzgus) [inline core/templates/layout.base.php]: I don't know, for me both version work faster.
But: [HTML5 does not require](https://www.w3.org/TR/html5/scripting-1.html#attr-script-type) the `type` attribute. My intention was to prevent current or future browsers triggering any kind of legacy or compatibility mode which **might** interrupt the concurrent resource loading.

In short: The change is _possibly_ not required for defeating the performance problem...

**Comment by ChristophWurst** [comment]
Wow! This is amazing. The page loads significantly faster. Thank you very much 😍

File: core/templates/layout.base.php [hunk 2]
```diff
@@ -18,15 +18,15 @@
 		<?php foreach($_['printcssfiles'] as $cssfile): ?>
 			<link rel="stylesheet" href="<?php print_unescaped($cssfile); ?>" media="print">
 		<?php endforeach; ?>
+		<?php foreach ($_['jsfiles'] as $jsfile): ?>
```
Comment (ChristophWurst) [inline core/templates/layout.base.php]: do we really have to move those lines up? This could potentially change/break the loading order with the scripts below (inline_ocjs)

File: core/templates/layout.guest.php [hunk 3]
```diff
@@ -19,15 +19,15 @@
 		<?php foreach($_['printcssfiles'] as $cssfile): ?>
 			<link rel="stylesheet" href="<?php print_unescaped($cssfile); ?>" media="print">
 		<?php endforeach; ?>
+		<?php foreach($_['jsfiles'] as $jsfile): ?>
```
Comment (ChristophWurst) [inline core/templates/layout.guest.php]: same problem/question here

File: core/templates/layout.user.php [hunk 4]
```diff
@@ -26,15 +26,15 @@
 		<?php foreach($_['printcssfiles'] as $cssfile): ?>
 			<link rel="stylesheet" href="<?php print_unescaped($cssfile); ?>" media="print">
 		<?php endforeach; ?>
+		<?php foreach($_['jsfiles'] as $jsfile): ?>
```
Comment (ChristophWurst) [inline core/templates/layout.user.php]: same

File: lib/private/legacy/template.php [hunk 5]
```diff
@@ -233,6 +233,9 @@ public function fetchPage($additionalParams = null) {
 			$headers = '';
 			foreach(OC_Util::$headers as $header) {
 				$headers .= '<'.\OCP\Util::sanitizeHTML($header['tag']);
+				if ( strcasecmp($header['tag'], 'script') == 0 and in_array('src', array_map('strtolower', array_keys($header['attributes']))) ) {
```
Comment (ChristophWurst) [inline lib/private/legacy/template.php]: use `===` for comparison and `&&` as logical `and`

**Review by ChristophWurst — DISMISSED** [review]

File: core/templates/layout.base.php (hunk 2 above)
Comment (michaelletzgus) [inline core/templates/layout.base.php]: Good point, it chnged the order. inline_ocjs will be executed first (while parsing), the deferred scripts after parsing but before the `DOMContentLoaded` event comes up.

The standard states: "The defer attribute is only for external scripts (should only be used if the src attribute is present)."

I have found no code in the  inline_ocjs which depends on other.
If the old order of execution (first inline, then extern) is desired the inline script code must become external. So why do we have internal code here at all? I think all these variables could be loaded form extern as well.

File: lib/private/legacy/template.php (hunk 5 above)
Comment (michaelletzgus) [inline lib/private/legacy/template.php]: ACK!

File: core/templates/layout.base.php (hunk 2 above)
Comment (michaelletzgus) [inline core/templates/layout.base.php]: @ChristophWurst 
In fact this does NOT change the order of execution when using `defer`for the external script, it only allows Firefox to download css and js in parallel.
With the inline_js block placed between css and js downloading of css and js happens synchronously. This seems to be a "feature" of at least Firefox...

Comment (ChristophWurst) [inline core/templates/layout.base.php]: I see, thanks for the explanation :-)

**Comment by benediktg** [comment]
I would recommend to eliminate the commits e3a4e4c8e864457294637c0bea0fe561901ee120, 49b6c4e80af9c3b1caa57b638ccb1d34c28ce271 and 9207852bc76ba6766b07a395d0eda0274c75307d because I think it's nonsense to merge `master` into this feature branch. ;)

**Comment by MorrisJobke** [comment]
> I would recommend to eliminate the commits e3a4e4c, 49b6c4e and 9207852 because I think it's nonsense to merge master into this feature branch. ;)

@michaelletzgus A rebase will automatically remove them ;)

**Comment by nickvergessen** [comment]
@michaelletzgus you don't need to close your PR, you can do a rebase easily on the same PR:

```
git checkout master
git pull nextcloud master
git checkout deferred-script-loading-v2
git rebase -i nextcloud/master
git push -f origin deferred-script-loading-v2
```

**Comment by michaelletzgus** [comment]
Sorry for the confusion, but git will definitively not my friend...

**Comment by nickvergessen** [comment]
Sorry for that, thanks anyway for trying and it seems to have worked 🎉

**Comment by michaelletzgus** [comment]
@nickvergessen No problem, maybe git will be my friend in a few days... ;)

**Comment by MorrisJobke** [comment]
> @nickvergessen No problem, maybe git will be my friend in a few days... ;)

@michaelletzgus Give https://try.github.io a try 😉 It helps to understand the concepts :)

**Review by MorrisJobke — APPROVED** [review]
Tested in Safari, Chrome, Firefox and IE11 and still works 👍

**Review by benediktg — APPROVED** [review]
I'm using this patch since a few days _(edit: with Firefox and Chromium)_ and the only issue I have noticed so far is that the code signature check fails. 😉

**Comment by MorrisJobke** [comment]
The only missing thing for this PR: the sign-off is missing: https://github.com/nextcloud/server/blob/master/CONTRIBUTING.md#sign-your-work

Then this is good to go in.

**Comment by michaelletzgus** [comment]
Signed!

**Comment by MorrisJobke** [comment]
We would really like to get some feedback from Firefox devs: https://bugzilla.mozilla.org/show_bug.cgi?id=1342781

Could someone please post there a link to our demo.nextcloud.com for testing this.

**Comment by xdmx** [comment]
> Could someone please post there a link to our demo.nextcloud.com for testing this.

Done

**Comment by DevSidious** [comment]
These changes broke my layout on Nextcloud 11.0.2 (stable).

Issues:

a) An artifact appeared in the upper left corner over the Nextcloud logo (white weird rectangle shape).
b) The avatar in the upper right corner got displaced (pushed downwards, under the header).
c) The small triangle indicating the presence of a menu at the right side of the avatar disappeared. 
d) The search icon at the left side of the avatar disappeared too.

In the case of c) and d) by "disappeared" I mean they are not visible. Clicking "around there" shows they exist but are not visible anymore.

Applications activated on my instance:

- Files
- Activity
- Gallery
- Calendar
- Passwords
- Notes
- Bookmarks
- Activity Log
- Audio player
- Tasks

Tested with and without this patch, on Chrome (56.0.2924.87 (64-bit)) and Mozilla (51.0.1 (32-bit)). 

As soon as I revert back to the original files provided with my Nextcloud download, the layout problems disappear.

ENVIRONMENT:

cat /etc/redhat-release
CentOS Linux release 7.3.1611 (Core)

httpd -v
Server version: Apache/2.4.25 (centos)
Server built:   Mar  2 2017 23:29:37

mysql -V
mysql  Ver 15.1 Distrib 10.1.21-MariaDB, for Linux (x86_64) using readline 5.1

php -v
PHP 7.1.2 (cli) (built: Feb 18 2017 11:22:37) ( NTS )
Copyright (c) 1997-2017 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2017 Zend Technologies

**Comment by michaelletzgus** [comment]
@DevSidious 
Is your testing instance public (or with http auth) available? This would make testing much easier.

**Comment by DevSidious** [comment]
It's my live personal setup, however I'm willing to help with the testing. Just let me know what I can do to help.

Looking forward to see this patch "production ready" so to speak. Great work by the way.

**Comment by michaelletzgus** [comment]
No errors is not good...
Hm... did you apply the patches "by hand" or did you use the patch tool?
Maybe you could post your modified version of:

/core/templates/layout.base.php
/core/templates/layout.group.php
/core/templates/layout.user.php
/lib/private/legacy/template.php

**Comment by DevSidious** [comment]
@michaelletzgus 

I went "by hand". 

1. Downloaded the raw files from here: https://github.com/nextcloud/server/pull/3696/files
2. Uploaded the files in the appropriate locations on my server overwriting the original ones;
3. Checked the owner for every uploaded file;
4. Changed the owner to apache:apache for every file;
5. Restarted httpd;

Right now my instance has been reverted to it's original state, but I can redo the whole thing and eventually take screenshots and paste file content if needed.

**Comment by DevSidious** [comment]
@michaelletzgus 

And you mean 

/core/templates/layout.**guest**.php and not /core/templates/layout.group.php , right ?

I have no file /core/templates/layout.group.php in my installation.

**Comment by DevSidious** [comment]
@michaelletzgus 

I redid the procedure. Screenshot:

<a href="http://3.1m.yt/_zwZ_ZE.jpg" target="_blank">ImgLand.net image</a>

**Comment by korelstar** [comment]
@DevSidious

The file you've downloaded is the developer version from Nextcloud 12. You should not mix it with your installation of Nextcloud 11.

Please revert to the original files from your installation and then apply the patch either by using `patch` or changing the respective lines manually.

**Comment by michaelletzgus** [comment]
@DevSidious @korelstar 
I can exactly reproduce this observation by placing latest (12.) and patched with `defer` version of the template files into a 11.0.2 installation - but NOT /lib/private/legacy/template.php.
If I replace /lib/private/legacy/template.php as well the layout is totaly screwd up.

So please apply the 4 patches to the v11.0.2 an try again. :-)

And: Thanks for testing!

**Comment by michaelletzgus** [comment]
@MorrisJobke 
Even if Chrome (and maybe Firefox in the future) are able to do parallel loading without `defer` we should add it anyway. Without `defer` the actual implicit behavior is up to the browser - but with `defer`we make it explicit.
Nevertheless, it's interesting to know why Chrome performs better than Firefox.

**Comment by DevSidious** [comment]
@michaelletzgus @korelstar 

Thanks for the help guys.

Attached the files for 11.0.2 stable with the current patch applied in case anyone else needs them. Tested on my live Nextcloud, instance, everything looks good. 

NOTE: 
I added the .txt extension because GitHub appears to be unable to recognize a zip archive as such... :|  

[nexcloud_11.0.2_stable_deferred_script_loading_v2_3696.zip.txt](https://github.com/nextcloud/server/files/836642/nexcloud_11.0.2_stable_deferred_script_loading_v2_3696.zip.txt)

**Comment by mm0** [comment]
Can someone explain what caused this to break? I have an installation of 11.0.1 that doesn't have the same behavior as 11.0.2.  Somehow 11.0.1 did not need the defer tag, yet it loads quickly.

**Comment by michaelletzgus** [comment]
I observed no difference between 11.0.1 and 11.0.2 when using Firefox, both are "slow" without `defer`.

**Comment by michaelletzgus** [comment]
@nickvergessen @MorrisJobke 
What to do next? Waiting for the Firefox people?

**Comment by nickvergessen** [comment]
When it doesn't get slower than it is at the moment, I'm fine with merging it

**Comment by Bubu** [comment]
Tested and works, applied on top of 11.0.2.

This is marked for version 12, any chance this can be applied for 11.0.3? It is really a difference of night and day with and without this patch.

**Comment by LukasReschke** [comment]
> This is marked for version 12, any chance this can be applied for 11.0.3? It is really a difference of night and day with and without this patch.

No. We do only backport critical patches. This here has the potential to breaking stuff and I'm still keen on feedback at https://bugzilla.mozilla.org/show_bug.cgi?id=1342781

**Comment by michaelletzgus** [comment]
I've found the first problem with `defer` in app files_markdown, editor.js.
I think this is not specific to this app but might be a more fundamental problem, maybe with jQuery, see below.
jQuery .then() in line 18 is fired to early:

```
   $.when(
                OCA.Files_Markdown.Preview.loadMarked(),
                OCA.Files_Markdown.Preview.loadHighlight(),
                OCA.Files_Markdown.Preview.loadKaTeX()
        ).then(function () {
                this.renderer = new marked.Renderer();
```
This results in "ReferenceError: marked is not defined".

Problem seems to be related to this:
https://github.com/jquery/jquery/issues/3271

**Comment by xdmx** [comment]
I've applied these changes to my nextcloud 11.0.0 and the difference of loading is a lot.

After few days of usage I've noticed that my external calendars (for example _webcal://www.calendarlabs.com/templates/ical/International-Holidays.ics_) aren't loading anymore.

Checking the console it says:
```
Content Security Policy: The page’s settings blocked the loading of a resource at self (“script-src 'nonce-Qug1WTdoeFpiZGgzS1g3cGxpMmNnV5B0WU15c05pdUI0eGZGODBEcEhLbz06Y1NZcHZrMHFYK3NWUWoraDhoVDF8MXVyRXZUNmVIelqxbnlIdXdxTlR2OD0=' 'unsafe-eval'”). Source: ;(function installGlobalHook(window) {
 ....
```

I haven't tried this under the v11.0.2

**Comment by nickvergessen** [comment]
Works fine here on 12. (with the very same url `webcal://www.calendarlabs.com/templates/ical/International-Holidays.ics`)
So I'd say either try 12 and open a new issue when it still happens there.

**Review by LukasReschke — CHANGES_REQUESTED** [review]
Just commenting here so that nobody merges this. I don't feel entirely confident here.

1. This breaks some apps as some people pointed out.
2. I still am curious about a feedback at https://bugzilla.mozilla.org/show_bug.cgi?id=1342781

I think it's better to go with https://github.com/nextcloud/server/pull/3988 instead. This achieves probably the same without too much dark magic. :)

**Comment by LukasReschke** [comment]
Replaced by https://github.com/nextcloud/server/pull/3988

**Comment by ChristophWurst** [comment]
Sure we saved some requests by combining js, but still FF loads stuff pretty slowly, which this PR actually addressed nicely. Can we re-open this and wait for FF dev's response? 🙏

**Comment by benediktg** [comment]
I agree – combining CSS and JS is rather a workaround for HTTP/1.1. And since every current version of nginx and Apache provides HTTP/2 support it's (IMHO) the Linux distributions' issue to support that and not Nextcloud's. ;)

Hence I would advise to not deny this PR.

**Comment by ajs124** [comment]
Actually, I use HTTP/2 and firefox still takes half a minute to load the files app on my firefox instance, so that's not really a fix.

**Comment by benediktg** [comment]
@ajs124 Could also be a configuration or setup issue.

**Comment by nsg** [comment]
I should point out that HAProxy do not support HTTP/2 (maybe 6-12 months away, or more), with is a really popular proxy software to load balance if you have a larger site (or, special needs). It's is also likely that a larger cooperation will be stuck with proprietary load balances that only support HTTP/1 for years. So support for HTTP/1 is relevant still.

I did some dark magic to enable HTTP/2 and it was faster, but still, it takes me 15-20 seconds on an empty cache and that's not a good user experience. Nextcloud feels "heavy and slow", not the experience we like to expose to users. It makes it harder to sell the idea to replace proprietary solutions with Nextcloud if feels really slow.

Still, I agree that this can't be merged if applications breaks, that is a worse user experience. If I understood this correctly most breakage is because libraries, like JQuery, are buggy with defer. I assume that will be fixed in the future.

In the future I think defer should be enabled (even if nice things like https://github.com/nextcloud/server/pull/3988 succeeds to limit the requests). Defer is a standard way to do things and it will speed up the user expertise for everyone. Maybe enable it by default in Nextcloud 13, 14 or 15,  ...

**Suggestion**
Make this a optional feature, disabled by default. With an appropriate warning message that this is a experimental feature that may break applications. This has the advantage that application developers may be motivated to support defer.

**Comment by Glandos** [comment]
I applied this a few seconds ago, on my NextCloud 11.0.1 behind HTTP/2 server.
There's no way I revert it. It just works for me, and is incredibly fast. Thanks a lot.

I know, this is a useless comment, but I think congrats should be written sometimes.

**Comment by michaelletzgus** [comment]
I'm still using it on my productive system.
There seems to be slight problem with the current master version - but at a first glance this might be fixable relatively easy.

**Comment by michaelletzgus** [comment]
No problems in current master anymore. :-)

**Comment by michaelletzgus** [comment]
Update: Still worling with almost all apps enabled. :)

**Comment by ViViDboarder** [comment]
Even with concatenation, this would provide a speed improvement by allowing the browser to load the rest of the images and markup while downloading the single script file. Concatenation is definitely welcome, but deferring loading seems like a good thing to have.

If the worry is making things `defer` by default, what about adding a parameter to `addScript` that allows `defer` or `async` to be provided? This would allow even better optimization for things that one knows can be done done `async` and safer implementation of `defer`.

**Comment by michaelletzgus** [comment]
Still fully operational with current master and all apps except ownBackup - but ownBackup does not work either.
@ViViDboarder Good point, I'll try this...

**Review by ghost — COMMENTED** [review]
Does anyone know if this applies to version 12?  Or do these only apply to version 11 of Nextcloud server?

**Comment by benediktg** [comment]
@lawtechsupport See #4854, in Nextcloud 13 there is a patch included which lets you enable deferred script loading via `'deferred_script_loading' => true,` in `config.php`. :wink:

But you also can have it earlier:
```sh
wget https://patch-diff.githubusercontent.com/raw/nextcloud/server/pull/4854.patch -O/tmp/deferred-script-loading.patch
cd /your/nextcloud/document/root
patch -p1 < /tmp/deferred-script-loading.patch
```
